cirq-google/cirq_google/engine/qcs_notebook.py

```diff
@@ -57,68 +55,64 @@ def get_qcs_objects_for_notebook(
-            google_cloud_signin_failed = True
-    else:  # pragma: no cover
-        os.environ['GOOGLE_CLOUD_PROJECT'] = project_id
-
-        # Following code runs the user through the Colab OAuth process.
-
-        # Checks for Google Application Default Credentials and runs
-        # interactive login if the notebook is executed in Colab. In
-        # case the notebook is executed in Jupyter notebook or other
-        # IPython runtimes, no interactive login is provided, it is
-        # assumed that the `GOOGLE_APPLICATION_CREDENTIALS` env var is
-        # set or `gcloud auth application-default login` was executed
-        # already. For more information on using Application Default Credentials
-        # see https://cloud.google.com/docs/authentication/production
-
+    # Check for Google Application Default Credentials and run
+    # interactive login if the notebook is executed in Colab. In
+    # case the notebook is executed in Jupyter notebook or other
+    # IPython runtimes, no interactive login is provided, it is
+    # assumed that the `GOOGLE_APPLICATION_CREDENTIALS` env var is
+    # set or `gcloud auth application-default login` was executed
+    # already. For more information on using Application Default Credentials
+    # see https://cloud.google.com/docs/authentication/production
+    try:
+        from IPython import get_ipython
+
+        in_colab = 'google.colab' in str(get_ipython())
+    except ImportError:
         in_colab = False
-        try:
-            from IPython import get_ipython
-
-            in_colab = 'google.colab' in str(get_ipython())
-
-            if in_colab:
-                from google.colab import auth
 
-                print("Getting OAuth2 credentials.")
-                print("Press enter after entering the verification code.")
-                auth.authenticate_user(clear_output=False)
-                print("Authentication complete.")
-            else:
-                print(
-                    "Notebook isn't executed with Colab, assuming "
-                    "Application Default Credentials are setup."
-                )
-        except:
-            pass
+    if in_colab:
+        from google.colab import auth
 
-        # End of Google Colab Authentication segment
-
-    device: cirq.Device
+        print("Getting OAuth2 credentials.")
+        print("Press enter after entering the verification code.")
+        try:
+            auth.authenticate_user(clear_output=False)
+            print("Authentication complete.")
+        except Exception as exc:
+            print(f"Authentication failed: {exc}")
+    else:
+        print("Not running in a colab kernel. Will use Application Default Credentials.")
+
+    # Attempt to connect to the Quantum Engine API, and use a simulator if unable to connect.
     sampler: Union[PhasedFSimEngineSimulator, QuantumEngineSampler]
-    if google_cloud_signin_failed or processor_id is None:
+    try:
+        engine = get_engine(project_id)
```

Comment:
I assume that this call + the try-except fulfills the role of the old `if project_id is None` check?

Reply:
It does that and more. Previously if you didn't specify a `project_id` the code would look for a `GOOGLE_CLOUD_PROJECT` environment variable, and if that is not defined would not connect to quantum engine. With the new code, we delegate to `google.auth.default()` which `get_engine` uses internally in the gapic machinery to get default credentials. This will still pull from `GOOGLE_CLOUD_PROJECT` if it is defined but will also respect other standard mechanisms of setting the project like if the user has used `gcloud config set project`. This is mostly an issue when connecting to a local jupyter backend, rather than running in hosted colab.